cmd/virt-tail/main.go

```diff
@@ -0,0 +1,181 @@
+func tailLogs() error {
+	t, err := tail.TailFile(gLogFile, tail.Config{
+		Follow:        true,
+		CompleteLines: true,
+		MustExist:     false,
+		ReOpen:        true,
+		Logger:        tail.DiscardingLogger,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		serr := t.Stop()
+		if serr != nil {
+			log.Log.Errorf("tail error: %v", serr)
+		}
+		t.Cleanup()
+	}()
+
+	for {
+		select {
+		case line, ok := <-t.Lines:
+			if !ok {
+				return errors.New("tail error")
+			}
+			if line.Err != nil {
+				return line.Err
+			}
+			fmt.Println(line.Text)
+		case <-gctx.Done():
+			return gctx.Err()
+		}
+	}
+}
+
+func checkFile(socketFile string, fileExists bool) bool {
+	if !fileExists {
+		_, err := os.Stat(socketFile)
+		return !os.IsNotExist(err)
+	}
+	return true
+}
+
+func watchFS() error {
+	socketFile := strings.TrimSuffix(gLogFile, "-log")
+	termFile := gLogFile + "-sigTerm"
+	socketExists := checkFile(socketFile, false)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	// Add a path.
+	err = watcher.Add(filepath.Dir(gLogFile))
+	if err != nil {
+		return err
+	}
+
+	// initial timeout for serial console socket creation
+	socketCheckCh := make(chan int)
+	time.AfterFunc(initialSocketTimeout, func() {
+		socketCheckCh <- 1
+	})
+
+	// Start listening for events.
+	for {
+		select {
+		case <-socketCheckCh:
+			if !checkFile(socketFile, socketExists) {
+				return errors.New("socketFile is still not ready")
+			} else {
+				socketExists = true
+			}
+			if checkFile(termFile, false) {
+				return errors.New("termFile was already there")
+			}
+		case event := <-watcher.Events:
+			if event.Has(fsnotify.Create) {
+				if event.Name == termFile {
+					// termination file got created, we should quickly terminate
+					return errors.New("termFile got detected")
+				} else if event.Name == socketFile {
+					// socket file got created
+					socketExists = true
+				}
+			} else if event.Has(fsnotify.Remove) {
+				if event.Name == socketFile {
+					// socket file got deleted, we should quickly terminate
+					return errors.New("socketFile got removed")
+				}
+			}
+		case werr := <-watcher.Errors:
+			return werr
+		case <-gctx.Done():
+			return gctx.Err()
+		}
+	}
+}
+
+func main() {
+	log.InitializeLogging("virt-tail")
+
+	logFile := pflag.String("logfile", "", "parth of the logfile to be streamed")
+
+	pflag.Parse()
+
+	if logFile == nil || *logFile == "" {
+		log.Log.Errorf("logfile flags must be provided")
+		os.Exit(1)
+	}
+	gLogFile = *logFile
+
+	// Create context that listens for the interrupt signal from the container runtime.
+	ctx, _ := signal.NotifyContext(context.Background(), os.Interrupt)
+
+	g, gctx = errgroup.WithContext(ctx)
+	g.Go(tailLogs)
+	g.Go(watchFS)
+
+	// wait for all errgroup goroutines
+	err := g.Wait()
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			os.Exit(0)
+		} else {
+			log.Log.Errorf("received error: %v", err)
+			os.Exit(1)
```

Comment:
In case of the termfile existing or the log file going away, should this always exit with error 1? I suppose when the termfile exists it should exit clean?

Reply:
yes, good catch